File: crm/app/internal/adapters/inbound/web/screen_tasks_board.go
// Package web — S07 Tasks Board screen handler.
// Inbound adapter: reads/writes via TaskQuerier, TaskWriter, TaskCreator.
// No business logic here — delegates to application.TaskService.
package web

import (
	"log"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/adapters/inbound/web/templates"
	"github.com/vantt/data-integration/crm/app/internal/domain"
)

func init() {
	register(func(r chi.Router, d *Deps) {
		// S07 — Tasks board (full page).
		r.Get("/tasks", d.handleTasksBoard)

		// M05 — Create task modal + submit.
		r.Get("/tasks/modal/create", d.handleModalCreateTask)
		r.Post("/tasks", d.handleCreateTask)

		// Task status transition (HTMX PATCH from board card).
		r.Patch("/tasks/{taskID}/status", d.handleTaskStatusTransition)

		// Generate tasks from action queue.
		r.Post("/tasks/generate", d.handleGenerateTasksFromQueue)
	})
}

// handleTasksBoard serves GET /tasks — S07 full page.
func (d *Deps) handleTasksBoard(w http.ResponseWriter, r *http.Request) {
	assignee := r.URL.Query().Get("assignee")
	statusFilter := r.URL.Query().Get("status")

	// When no explicit status filter, show all statuses for the board.
	tasks, err := d.Tasks.ListTasks(r.Context(), assignee, statusFilter)
	if err != nil {
		log.Printf("tasks board: list: %v", err)
		tasks = []domain.Task{}
	}

	// Group by status for the kanban columns.
	grouped := groupTasksByStatus(tasks)

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.TasksBoardPage(grouped, assignee, statusFilter).Render(r.Context(), w); err != nil {
		log.Printf("tasks board: render: %v", err)
	}
}

// handleModalCreateTask serves GET /tasks/modal/create — M05 empty form fragment.
func (d *Deps) handleModalCreateTask(w http.ResponseWriter, r *http.Request) {
	partyID := r.URL.Query().Get("party_id") // optional pre-fill from Customer 360
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ModalCreateTask(partyID).Render(r.Context(), w); err != nil {
		log.Printf("modal create task: render: %v", err)
	}
}

// handleCreateTask serves POST /tasks — M05 submit; creates a new task.
func (d *Deps) handleCreateTask(w http.ResponseWriter, r *http.Request) {
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}

	title := strings.TrimSpace(r.FormValue("title"))
	if title == "" {
		http.Error(w, "title required", http.StatusBadRequest)
		return
	}

	t := &domain.Task{
		Title:       title,
		Description: strings.TrimSpace(r.FormValue("description")),
		Status:      "open",
		Source:      "manual",
	}

	// Optional fields.
	if dueAt := strings.TrimSpace(r.FormValue("due_at")); dueAt != "" {
		t.DueAt = &dueAt
	}
	if partyID := strings.TrimSpace(r.FormValue("party_id")); partyID != "" {
		t.PartyID = &partyID
	}
	if assignee := strings.TrimSpace(r.FormValue("assignee_user_id")); assignee != "" {
		t.AssigneeUserID = &assignee
	}

	if err := d.TaskCreator.CreateTask(r.Context(), t); err != nil {
		log.Printf("create task: %v", err)
		http.Error(w, "failed to create task", http.StatusInternalServerError)
		return
	}

	// Close modal and redirect to board to show new task.
	w.Header().Set("HX-Trigger", `{"closeModal":true}`)
	w.Header().Set("HX-Redirect", "/tasks")
	w.WriteHeader(http.StatusOK)
}

// handleTaskStatusTransition serves PATCH /tasks/{taskID}/status — board card status update.
func (d *Deps) handleTaskStatusTransition(w http.ResponseWriter, r *http.Request) {
	taskID := chi.URLParam(r, "taskID")
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	newStatus := strings.TrimSpace(r.FormValue("status"))
	if newStatus == "" {
		http.Error(w, "status required", http.StatusBadRequest)
		return
	}
	if err := d.TaskWriter.TransitionStatus(r.Context(), taskID, newStatus); err != nil {
		log.Printf("task status transition %s→%s: %v", taskID, newStatus, err)
		http.Error(w, "failed to transition", http.StatusInternalServerError)
		return
	}
	// Refresh the whole board so the card moves column.
	w.Header().Set("HX-Redirect", "/tasks")
	w.WriteHeader(http.StatusOK)
}

// handleGenerateTasksFromQueue serves POST /tasks/generate — creates tasks from action queue.
// Returns a 200 with a count fragment; graceful-empty if cache absent (returns 0, no error).
func (d *Deps) handleGenerateTasksFromQueue(w http.ResponseWriter, r *http.Request) {
	if d.TaskGen == nil {
		// TaskGen not wired (e.g. in test fixtures); redirect gracefully.
		w.Header().Set("HX-Redirect", "/tasks")
		w.WriteHeader(http.StatusOK)
		return
	}
	count, err := d.TaskGen.GenerateTasksFromActionQueue(r.Context())
	if err != nil {
		log.Printf("generate tasks from queue: %v", err)
		// Non-fatal: still redirect to board.
	}
	log.Printf("generate tasks from queue: created %d tasks", count)
	// Redirect to board so the new tasks show up.
	w.Header().Set("HX-Redirect", "/tasks")
	w.WriteHeader(http.StatusOK)
}

// ── local helpers ─────────────────────────────────────────────────────────────

// groupTasksByStatus partitions tasks into open/doing/done/cancelled buckets.
func groupTasksByStatus(tasks []domain.Task) map[string][]domain.Task {
	out := map[string][]domain.Task{
		"open":      {},
		"doing":     {},
		"done":      {},
		"cancelled": {},
	}
	for _, t := range tasks {
		bucket := t.Status
		if _, ok := out[bucket]; !ok {
			bucket = "open"
		}
		out[bucket] = append(out[bucket], t)
	}
	return out
}

File: crm/src/adapters/inbound/web/screen_tasks_board.py
"""Web adapter — S07 Tasks Board screen.

FastAPI router mirroring screen_tasks_board.go.
Serves full Kanban HTML page + HTMX status-transition endpoint.
No business logic — thin adapter only.
"""
from __future__ import annotations

import logging
from typing import Optional, Protocol

from fastapi import APIRouter, Form, Request, Response
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates

from domain.entities.task import Task, VALID_TASK_STATUSES

log = logging.getLogger(__name__)

# ── Service protocols ─────────────────────────────────────────────────────────

class TaskQuerier(Protocol):
    def list_tasks(self, assignee: str, status: str) -> list[Task]: ...


class UserQuerier(Protocol):
    def list_active(self) -> list: ...


class TaskWriter(Protocol):
    def transition_status(self, task_id: str, new_status: str) -> None: ...


class TaskCreator(Protocol):
    def create_task(self, task_data: dict) -> object: ...


class TaskGenerator(Protocol):
    def generate_tasks_from_action_queue(self) -> int: ...


# ── Helpers ───────────────────────────────────────────────────────────────────

def _group_by_status(tasks: list[Task]) -> dict[str, list[Task]]:
    """Partition tasks into kanban columns: open / doing / done / cancelled."""
    grouped: dict[str, list[Task]] = {
        "open": [], "doing": [], "done": [], "cancelled": [],
    }
    for t in tasks:
        bucket = t.status if t.status in grouped else "open"
        grouped[bucket].append(t)
    return grouped


def _task_status_css(status: str) -> str:
    return {
        "open": "bdg--good",
        "doing": "bdg--warn",
        "done": "",
        "cancelled": "bdg--off",
    }.get(status, "")


# ── Router factory ────────────────────────────────────────────────────────────

def make_tasks_board_router(
    templates: Jinja2Templates,
    task_querier: TaskQuerier,
    task_writer: TaskWriter,
    task_creator: TaskCreator,
    task_generator: Optional[TaskGenerator] = None,
    user_querier: Optional[UserQuerier] = None,
) -> APIRouter:
    """Return APIRouter wired with all Tasks Board routes."""
    router = APIRouter()

    @router.get("/tasks", response_class=HTMLResponse)
    async def handle_tasks_board(request: Request) -> Response:
        assignee = request.query_params.get("assignee", "")
        status_filter = request.query_params.get("status", "")
        try:
            tasks = task_querier.list_tasks(assignee, status_filter)
        except Exception as exc:
            log.error("tasks board: list: %s", exc)
            tasks = []
        users = user_querier.list_active() if user_querier else []
        grouped = _group_by_status(tasks)
        return templates.TemplateResponse(
            "tasks_board.html",
            {
                "request": request,
                "grouped": grouped,
                "assignee_filter": assignee,
                "status_filter": status_filter,
                "task_status_css": _task_status_css,
                "users": users,
            },
        )

    @router.get("/tasks/modal/create", response_class=HTMLResponse)
    async def handle_modal_create_task(request: Request) -> Response:
        party_id = request.query_params.get("party_id", "")
        return templates.TemplateResponse(
            "fragments/modal_create_task.html",
            {"request": request, "party_id": party_id},
        )

    @router.post("/tasks", response_class=HTMLResponse)
    async def handle_create_task(
        request: Request,
        title: str = Form(...),
        description: str = Form(default=""),
        due_at: str = Form(default=""),
        party_id: str = Form(default=""),
        assignee_user_id: str = Form(default=""),
    ) -> Response:
        title = title.strip()
        if not title:
            return HTMLResponse("title required", status_code=400)
        current_user = getattr(request.state, "current_user", None)
        task_data = {
            "title": title,
            "description": description.strip() or None,
            "status": "open",
            "source": "manual",
            "priority": 0,
            "due_at": due_at.strip() or None,
            "party_id": party_id.strip() or None,
            "assignee_user_id": assignee_user_id.strip() or None,
            "created_by": current_user.user_id if current_user else None,
        }
        try:
            task_creator.create_task(task_data)
        except Exception as exc:
            log.error("create task: %s", exc)
            return HTMLResponse("failed to create task", status_code=500)
        return Response(
            status_code=200,
            headers={"HX-Trigger": '{"closeModal":true}', "HX-Redirect": "/tasks"},
        )

    @router.patch("/tasks/{task_id}/status", response_class=HTMLResponse)
    async def handle_task_status_transition(
        request: Request, task_id: str, status: str = Form(...)
    ) -> Response:
        status = status.strip()
        if status not in VALID_TASK_STATUSES:
            return HTMLResponse("invalid status", status_code=400)
        try:
            task_writer.transition_status(task_id, status)
        except Exception as exc:
            log.error("task status transition %s→%s: %s", task_id, status, exc)
            return HTMLResponse("failed to transition", status_code=500)
        return Response(status_code=200, headers={"HX-Redirect": "/tasks"})

    @router.post("/tasks/generate", response_class=HTMLResponse)
    async def handle_generate_tasks_from_queue(request: Request) -> Response:
        if task_generator is None:
            return Response(status_code=200, headers={"HX-Redirect": "/tasks"})
        try:
            count = task_generator.generate_tasks_from_action_queue()
            log.info("generate tasks from queue: created %d tasks", count)
        except Exception as exc:
            log.error("generate tasks from queue: %s", exc)
        return Response(status_code=200, headers={"HX-Redirect": "/tasks"})

    return router
